etcd failure should not impact data plane

#### Setup Details
CIS Version : 2.11.1
Build: f5networks/k8s-bigip-ctlr:2.11.1
BIGIP Version: Big IP 15.1.8
AS3 Version: 3.26
Agent Mode: AS3
Orchestration: K8S
Orchestration Version: 1.21.5
Pool Mode: Nodeport
Additional Setup details: Calico

#### Description  
CIS behavior is to delete configuration like pool members when etcd is down ( getEndpoints function fails) 

#### Steps To Reproduce
1) Deploy application e.g. with AS3 configmap
2) etcd becomes unreachable (e.g. due to disk failure/ full disk)
3) Confirm CIS removes pool members when discovery returns no endpoints due to getEndpoints failure

#### Expected Result 

Even if etcd is down, data plan should remain in tact as pods could still be in running state ready for processing traffic

#### Actual Result  

pool members deleted causing downtime

#### Diagnostic Information
```logs
2023/05/25 08:45:31 [ERROR] [CORE] Error getting service list. etcdserver: leader changed
2023/05/25 08:45:31 [DEBUG] [CORE] Discovered members for service my-service1-ns/my-service1 is []
```  

Below KB article outlines the same issue
https://my.f5.com/manage/s/article/K41481489

--- a/docs/RELEASE-NOTES.rst
+++ b/docs/RELEASE-NOTES.rst
@@ -20,8 +20,7 @@ Bug Fixes
 ````````````
 * `Issue 2941 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2941>`_: Fix for services with same name in different namespaces in NodePortLocal mode
 * `Issue 2850 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2850>`_: Fix for AS3 config updated every 30 seconds by CIS with default ingress backend
-
-
+* `Issue 2909 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2909>`_: Fix for empty pool members when K8S API server throws any error
 2.13.1
 -------------
 Bug Fixes

--- a/pkg/agent/as3/as3ConfigMap.go
+++ b/pkg/agent/as3/as3ConfigMap.go
@@ -21,6 +21,7 @@ const (
 func (am *AS3Manager) prepareResourceAS3ConfigMaps() (
 	[]*AS3ConfigMap,
 	string,
+	error,
 ) {
 	var as3Cfgmaps []*AS3ConfigMap
 	var overriderAS3CfgmapData string
@@ -64,7 +65,11 @@ func (am *AS3Manager) prepareResourceAS3ConfigMaps() (
 				}
 			}
 
-			tenantMap, endPoints := am.processCfgMap(rscCfgMap)
+			tenantMap, endPoints, err := am.processCfgMap(rscCfgMap)
+			// Skip processing further if error encountered while processing configMap
+			if err != nil {
+				return nil, "", err
+			}
 			if tenantMap == nil {
 				continue
 			}
@@ -90,13 +95,17 @@ func (am *AS3Manager) prepareResourceAS3ConfigMaps() (
 				Validated: true,
 			}
 			rscCfgMap.Data = am.getTenantObjects(tenants)
-			tenantMap, endPoints := am.processCfgMap(rscCfgMap)
+			tenantMap, endPoints, err := am.processCfgMap(rscCfgMap)
+			// Skip processing further if error encountered while processing configMap
+			if err != nil {
+				return nil, "", err
+			}
 			cfgmap.config = tenantMap
 			cfgmap.endPoints = endPoints
 			as3Cfgmaps = append(as3Cfgmaps, cfgmap)
 		}
 	}
-	return as3Cfgmaps, overriderAS3CfgmapData
+	return as3Cfgmaps, overriderAS3CfgmapData, nil
 }
 
 func (am *AS3Manager) isValidConfigmap(cfgmap *AgentCfgMap) (string, bool) {
@@ -127,6 +136,7 @@ func (am *AS3Manager) isValidConfigmap(cfgmap *AgentCfgMap) (string, bool) {
 func (am *AS3Manager) processCfgMap(rscCfgMap *AgentCfgMap) (
 	map[string]interface{},
 	[]Member,
+	error,
 ) {
 	as3Tmpl := as3Template(rscCfgMap.Data)
 	obj, ok := getAS3ObjectFromTemplate(as3Tmpl)
@@ -134,7 +144,7 @@ func (am *AS3Manager) processCfgMap(rscCfgMap *AgentCfgMap) (
 		log.Errorf("[AS3][Configmap] Error processing AS3 template")
 		log.Errorf("[AS3]Error in processing the ConfigMap: %v/%v",
 			rscCfgMap.Namespace, rscCfgMap.Name)
-		return nil, nil
+		return nil, nil, nil
 	}
 
 	if _, ok := obj[tenantName(DEFAULT_PARTITION)]; ok {
@@ -142,7 +152,7 @@ func (am *AS3Manager) processCfgMap(rscCfgMap *AgentCfgMap) (
 			rscCfgMap.Namespace, rscCfgMap.Name)
 		log.Errorf("[AS3] CIS managed partition <%s> should not be used in ConfigMaps as a Tenant",
 			DEFAULT_PARTITION)
-		return nil, nil
+		return nil, nil, nil
 	}
 
 	var tmp interface{}
@@ -150,7 +160,7 @@ func (am *AS3Manager) processCfgMap(rscCfgMap *AgentCfgMap) (
 	// unmarshall the template of type string to interface
 	err := json.Unmarshal([]byte(as3Tmpl), &tmp)
 	if nil != err {
-		return nil, nil
+		return nil, nil, nil
 	}
 
 	// convert tmp to map[string]interface{}, This conversion will help in traversing the as3 object
@@ -183,7 +193,11 @@ func (am *AS3Manager) processCfgMap(rscCfgMap *AgentCfgMap) (
 			appObj := tenantObj[string(app)].(map[string]interface{})
 			for _, pn := range pools {
 				poolObj := appObj[string(pn)].(map[string]interface{})
-				eps := rscCfgMap.GetEndpoints(am.getSelector(tnt, app, pn), rscCfgMap.Namespace)
+				eps, err := rscCfgMap.GetEndpoints(am.getSelector(tnt, app, pn), rscCfgMap.Namespace)
+				// If there is some error while fetching the endpoint from API server then skip processing further
+				if nil != err {
+					return nil, nil, err
+				}
 				// Handle an empty value
 				if len(eps) == 0 {
 					continue
@@ -236,7 +250,7 @@ func (am *AS3Manager) processCfgMap(rscCfgMap *AgentCfgMap) (
 		}
 		tenantMap[string(tnt)] = tenantObj
 	}
-	return tenantMap, members
+	return tenantMap, members, nil
 }
 
 // Method prepares and returns the label selector in string format

--- a/pkg/agent/as3/as3Manager.go
+++ b/pkg/agent/as3/as3Manager.go
@@ -220,7 +220,7 @@ func updateTenantMap(tempAS3Config AS3Config) AS3Config {
 	return tempAS3Config
 }
 
-func (am *AS3Manager) postAS3Declaration(rsReq ResourceRequest) (bool, string) {
+func (am *AS3Manager) postAS3Declaration(rsReq ResourceRequest) (bool, string, error) {
 	am.ResourceRequest = rsReq
 
 	// as3Config := am.as3ActiveConfig
@@ -231,14 +231,20 @@ func (am *AS3Manager) postAS3Declaration(rsReq ResourceRequest) (bool, string) {
 	// Process Route or Ingress
 	as3Config.resourceConfig = am.prepareAS3ResourceConfig()
 
+	var err error
 	// Process all Configmaps (including overrideAS3)
-	as3Config.configmaps, as3Config.overrideConfigmapData = am.prepareResourceAS3ConfigMaps()
-
+	as3Config.configmaps, as3Config.overrideConfigmapData, err = am.prepareResourceAS3ConfigMaps()
+	// Skip posting AS3 declaration if error encountered while processing configMap to avoid possible wrong declaration
+	// getting posted as the pool members may be empty if error is encountered while connecting with api server
+	if err != nil {
+		return false, "", err
+	}
 	if am.FilterTenants {
 		updateTenantMap(*as3Config)
 	}
 
-	return am.postAS3Config(*as3Config)
+	posted, url := am.postAS3Config(*as3Config)
+	return posted, url, nil
 }
 
 func (am *AS3Manager) getADC() map[string]interface{} {
@@ -561,7 +567,11 @@ func (am *AS3Manager) ConfigDeployer() {
 		case msgReq = <-am.ReqChan:
 		case <-time.After(1 * time.Microsecond):
 		}
-		posted, event := am.postAS3Declaration(msgReq.ResourceRequest)
+		posted, event, err := am.postAS3Declaration(msgReq.ResourceRequest)
+		// Skip processing further if error is encountered during preparing and posting of AS3 declaration
+		if err != nil {
+			continue
+		}
 		am.updateNetworkingConfig()
 
 		// To handle general errors
@@ -572,7 +582,11 @@ func (am *AS3Manager) ConfigDeployer() {
 				timeout = postDelayTimeout
 			}
 			log.Debugf("[AS3] Error handling for event %v", event)
-			posted, event = am.postOnEventOrTimeout(timeout)
+			var err error
+			posted, event, err = am.postOnEventOrTimeout(timeout)
+			if err != nil {
+				continue
+			}
 			am.updateNetworkingConfig()
 		}
 		firstPost = false
@@ -598,12 +612,13 @@ func (am *AS3Manager) failureHandler() (bool, string) {
 }
 
 // Helper method used by configDeployer to handle error responses received from BIG-IP
-func (am *AS3Manager) postOnEventOrTimeout(timeout time.Duration) (bool, string) {
+func (am *AS3Manager) postOnEventOrTimeout(timeout time.Duration) (bool, string, error) {
 	select {
 	case msgReq := <-am.ReqChan:
 		return am.postAS3Declaration(msgReq.ResourceRequest)
 	case <-time.After(timeout):
-		return am.failureHandler()
+		posted, url := am.failureHandler()
+		return posted, url, nil
 	}
 }
 

--- a/pkg/agent/as3/as3Manager_test.go
+++ b/pkg/agent/as3/as3Manager_test.go
@@ -35,9 +35,9 @@ func newMockAS3Manager(params *Params) *mockAS3Manager {
 	}
 }
 
-func mockGetEndPoints(string, string) []Member {
+func mockGetEndPoints(string, string) ([]Member, error) {
 	return []Member{{Address: "1.1.1.1", Port: 80, SvcPort: 80},
-		{Address: "2.2.2.2", Port: 80, SvcPort: 80}}
+		{Address: "2.2.2.2", Port: 80, SvcPort: 80}}, nil
 }
 
 func (m *mockAS3Manager) shutdown() error {
@@ -122,7 +122,7 @@ var _ = Describe("AS3Manager Tests", func() {
 				},
 			)
 			as3config := &AS3Config{}
-			as3config.configmaps, _ = mockMgr.prepareResourceAS3ConfigMaps()
+			as3config.configmaps, _, _ = mockMgr.prepareResourceAS3ConfigMaps()
 			result := mockMgr.getUnifiedDeclaration(as3config)
 			Expect(string(result)).To(MatchJSON(unified), "Failed to Create JSON with correct configuration")
 		})
@@ -158,7 +158,7 @@ var _ = Describe("AS3Manager Tests", func() {
 				},
 			)
 			as3config := &AS3Config{}
-			as3config.configmaps, _ = mockMgr.prepareResourceAS3ConfigMaps()
+			as3config.configmaps, _, _ = mockMgr.prepareResourceAS3ConfigMaps()
 
 			routeCfg := readConfigFile(configPath + "as3_route.json")
 			err := json.Unmarshal([]byte(routeCfg), &routeAdc)
@@ -250,7 +250,7 @@ var _ = Describe("AS3Manager Tests", func() {
 				},
 			)
 			as3config := &AS3Config{}
-			_, as3config.overrideConfigmapData = mockMgr.prepareResourceAS3ConfigMaps()
+			_, as3config.overrideConfigmapData, _ = mockMgr.prepareResourceAS3ConfigMaps()
 
 			routeCfg := readConfigFile(configPath + "as3_route.json")
 			err := json.Unmarshal([]byte(routeCfg), &routeAdc)
@@ -425,7 +425,7 @@ var _ = Describe("AS3Manager Tests", func() {
 				AgentCfgmaps: []*AgentCfgMap{},
 			}
 			mockMgr.ReqChan <- MessageRequest{ReqID: 1, MsgType: "test", ResourceRequest: resourceRequest}
-			result, output := mockMgr.postOnEventOrTimeout(timeoutSmall)
+			result, output, _ := mockMgr.postOnEventOrTimeout(timeoutSmall)
 			close(mockMgr.ReqChan)
 			Expect(result).To(BeTrue(), "Posting Failed")
 			Expect(output).To(Equal("statusOK"), "Posting Failed")
@@ -475,7 +475,7 @@ var _ = Describe("AS3Manager Tests", func() {
 				},
 			)
 			mockMgr.as3ActiveConfig.unifiedDeclaration = as3Declaration(readConfigFile(configPath + "as3config_multi_cm_unified.json"))
-			result, output := mockMgr.postAS3Declaration(resourceRequest)
+			result, output, _ := mockMgr.postAS3Declaration(resourceRequest)
 			Expect(result).To(BeTrue(), "Posting Failed")
 			Expect(output).To(Equal(responseStatusOk), "Posting Failed")
 			// Verify that tenant3 is deleted from active configMap

--- a/pkg/appmanager/appManager.go
+++ b/pkg/appmanager/appManager.go
@@ -20,6 +20,7 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"k8s.io/apimachinery/pkg/api/errors"
 	"net"
 	"reflect"
 	"sort"
@@ -1617,7 +1618,10 @@ func (appMgr *Manager) syncConfigMaps(
 			// A service can be considered as an as3 configmap associated service only when it has these 3 labels
 			if tntOk && appOk && poolOk {
 				//TODO: Sorting endpoints members
-				members := appMgr.getEndpoints(selector, sKey.Namespace)
+				members, err := appMgr.getEndpoints(selector, sKey.Namespace)
+				if err != nil {
+					return err
+				}
 				if _, ok := appMgr.agentCfgMapSvcCache[key]; !ok {
 					if len(members) != 0 {
 						appMgr.agentCfgMapSvcCache[key] = &SvcEndPointsCache{
@@ -1716,15 +1720,22 @@ func (appMgr *Manager) syncConfigMaps(
 					continue
 				}
 				// Check if profile is contained in a Secret
-				secret, err := appMgr.kubeClient.CoreV1().Secrets(cm.ObjectMeta.Namespace).
-					Get(context.TODO(), profile.Name, metav1.GetOptions{})
-				if err != nil {
+				appInf, found := appMgr.getNamespaceInformer(cm.ObjectMeta.Namespace)
+				if !found {
+					log.Debugf("[CORE] No Informer found while fetching secret with name '%s' in namespace '%s'",
+						profile.Name, sKey.Namespace)
+					return fmt.Errorf("[CORE] No Informer found while fetching secret with name '%s' in namespace '%s'",
+						profile.Name, sKey.Namespace)
+				}
+				obj, found, err := appInf.secretInformer.GetIndexer().GetByKey(
+					fmt.Sprintf("%s/%s", cm.ObjectMeta.Namespace, profile.Name))
+				if err != nil || !found {
 					// No secret, so we assume the profile is a BIG-IP default
 					log.Debugf("[CORE] No Secret with name '%s' in namespace '%s', "+
 						"parsing secretName as path instead.", profile.Name, sKey.Namespace)
 					continue
 				}
-
+				secret := obj.(*v1.Secret)
 				appMgr.rsrcSSLCtxt[profile.Name] = secret
 				err, updated := appMgr.createSecretSslProfile(rsCfg, secret)
 				if err != nil {
@@ -1738,9 +1749,12 @@ func (appMgr *Manager) syncConfigMaps(
 		}
 
 		rsName := rsCfg.GetNameRef()
-		ok, found, updated := appMgr.handleConfigForType(
+		ok, found, updated, err := appMgr.handleConfigForType(
 			rsCfg, sKey, rsMap, rsName, svcPortMap,
 			svc, appInf, []string{}, nil)
+		if err != nil {
+			return err
+		}
 		if !ok {
 			stats.vsUpdated += updated
 			continue
@@ -2219,7 +2233,7 @@ func (appMgr *Manager) syncRoutes(
 				rsCfg.SetPolicy(pol)
 			}
 
-			ok, found, updated := appMgr.handleConfigForType(
+			ok, found, updated, _ := appMgr.handleConfigForType(
 				rsCfg, sKey, rsMap, rsName, svcPortMap,
 				svc, appInf, svcNames, nil)
 			if ok {
@@ -2472,7 +2486,7 @@ func (appMgr *Manager) handleConfigForTypeIngress(
 						correctBackend, reason, msg = true, "", ""
 					}
 				} else if appMgr.poolMemberType == NodePortLocal {
-					correctBackend, reason, msg =
+					correctBackend, reason, msg, _ =
 						appMgr.updatePoolMembersForNPL(svc, svcKey, rsCfg, plIdx)
 				} else {
 					correctBackend, reason, msg =
@@ -2537,7 +2551,7 @@ func (appMgr *Manager) handleConfigForType(
 	appInf *appInformer,
 	currResourceSvcs []string, // Used for Ingress/Routes
 	obj interface{}, // Used for writing events
-) (bool, int, int) {
+) (bool, int, int, error) {
 	vsFound := 0
 	vsUpdated := 0
 
@@ -2555,7 +2569,7 @@ func (appMgr *Manager) handleConfigForType(
 		}
 	}
 	if !found {
-		return false, vsFound, vsUpdated
+		return false, vsFound, vsUpdated, nil
 	}
 
 	// Make sure pool members from the old config are applied to the new
@@ -2623,7 +2637,7 @@ func (appMgr *Manager) handleConfigForType(
 			msg := "Service " + pool.ServiceName + " has not been found."
 			appMgr.recordV1IngressEvent(obj.(*netv1.Ingress), "ServiceNotFound", msg)
 		}
-		return false, vsFound, vsUpdated
+		return false, vsFound, vsUpdated, nil
 	}
 
 	// Update pool members.
@@ -2648,8 +2662,12 @@ func (appMgr *Manager) handleConfigForType(
 				correctBackend, reason, msg = true, "", ""
 			}
 		} else if appMgr.poolMemberType == NodePortLocal {
-			correctBackend, reason, msg =
+			var err error
+			correctBackend, reason, msg, err =
 				appMgr.updatePoolMembersForNPL(svc, svcKey, rsCfg, plIdx)
+			if err != nil {
+				return false, 0, 0, err
+			}
 		} else {
 			correctBackend, reason, msg =
 				appMgr.updatePoolMembersForCluster(svc, svcKey, rsCfg, appInf, plIdx)
@@ -2675,10 +2693,10 @@ func (appMgr *Manager) handleConfigForType(
 		if !serviceMatch(currResourceSvcs, sKey) {
 			//pool found but service not matched with current resource backend. So endpoints are not updated for correct pool
 			//So keep the resource as not processed.
-			return false, vsFound, vsUpdated
+			return false, vsFound, vsUpdated, nil
 		}
 	}
-	return true, vsFound, vsUpdated
+	return true, vsFound, vsUpdated, nil
 }
 
 func (appMgr *Manager) syncPoolMembers(rsName NameRef, rsCfg *ResourceConfig) {
@@ -2737,10 +2755,13 @@ func (appMgr *Manager) updatePoolMembersForNPL(
 	svcKey ServiceKey,
 	rsCfg *ResourceConfig,
 	index int,
-) (bool, string, string) {
+) (bool, string, string, error) {
 	//get pods for service
 	if svc.Spec.Type == v1.ServiceTypeClusterIP || svc.Spec.Type == v1.ServiceTypeLoadBalancer {
-		pods := appMgr.GetPodsForService(svcKey.Namespace, svcKey.ServiceName)
+		pods, err := appMgr.GetPodsForService(svcKey.Namespace, svcKey.ServiceName)
+		if err != nil {
+			return false, "", "", err
+		}
 		if pods != nil {
 			for _, portSpec := range svc.Spec.Ports {
 				if portSpec.Port == svcKey.ServicePort {
@@ -2756,11 +2777,11 @@ func (appMgr *Manager) updatePoolMembersForNPL(
 		if rsCfg.Pools[index].Members == nil {
 			log.Debugf("[CORE]Endpoints could not be fetched for service %v with port %v", svcKey.ServiceName, svcKey.ServicePort)
 		}
-		return true, "", ""
+		return true, "", "", nil
 	} else {
 		msg := "[CORE] Requested service backend " + svcKey.ServiceName + " not of ClusterIP or LoadBalancer type supported for NPL"
 		log.Debug(msg)
-		return false, "IncorrectBackendServiceType", msg
+		return false, "IncorrectBackendServiceType", msg, nil
 	}
 }
 
@@ -3245,7 +3266,7 @@ func createLabel(label string) (labels.Selector, error) {
 // When controller is in ClusterIP mode, returns a list of Cluster IP Address and Service Port. Also, it accumulates
 // members for static ARP entry population.
 
-func (appMgr *Manager) getEndpoints(selector, namespace string) []Member {
+func (appMgr *Manager) getEndpoints(selector, namespace string) ([]Member, error) {
 	var members []Member
 	uniqueMembersMap := make(map[Member]struct{})
 
@@ -3267,7 +3288,7 @@ func (appMgr *Manager) getEndpoints(selector, namespace string) []Member {
 
 		if err != nil {
 			log.Errorf("[CORE] Error getting service list. %v", err)
-			return nil
+			return nil, err
 		}
 		svcItems = services.Items
 	} else {
@@ -3314,7 +3335,7 @@ func (appMgr *Manager) getEndpoints(selector, namespace string) []Member {
 				)
 				if err != nil {
 					log.Debugf("[CORE] Error getting endpoints for service %v", service.Name)
-					continue
+					return nil, err
 				}
 				if len(endpointsList.Items) == 0 {
 					log.Debugf("[CORE] Endpoints for service %v not found", service.Name)
@@ -3340,7 +3361,10 @@ func (appMgr *Manager) getEndpoints(selector, namespace string) []Member {
 				}
 			}
 		} else if appMgr.poolMemberType == NodePortLocal { // Controller is in NodePortLocal Mode
-			pods := appMgr.GetPodsForService(service.Namespace, service.Name)
+			pods, err := appMgr.GetPodsForService(service.Namespace, service.Name)
+			if err != nil {
+				return nil, err
+			}
 			if pods != nil {
 				for _, portSpec := range service.Spec.Ports {
 					podPort := portSpec.TargetPort
@@ -3369,7 +3393,7 @@ func (appMgr *Manager) getEndpoints(selector, namespace string) []Member {
 			}*/
 		}
 	}
-	return members
+	return members, nil
 }
 
 func (appMgr *Manager) exposeKubernetesService(
@@ -3468,12 +3492,12 @@ func (appMgr *Manager) GetServicesForPod(pod *v1.Pod) []*v1.Service {
 	return svcList
 }
 
-func (appMgr *Manager) GetPodsForService(namespace, serviceName string) []*v1.Pod {
+func (appMgr *Manager) GetPodsForService(namespace, serviceName string) ([]*v1.Pod, error) {
 	svcKey := namespace + "/" + serviceName
 	crInf, ok := appMgr.getNamespaceInformer(namespace)
 	if !ok && !appMgr.hubMode {
 		log.Errorf("Informer not found for namespace: %v", namespace)
-		return nil
+		return nil, nil
 	}
 	var svc interface{}
 	var err error
@@ -3484,28 +3508,31 @@ func (appMgr *Manager) GetPodsForService(namespace, serviceName string) []*v1.Po
 		svc, found, err = crInf.svcInformer.GetIndexer().GetByKey(svcKey)
 		if !found {
 			log.Errorf("Error: Service %v not found", svcKey)
-			return nil
+			return nil, nil
 		}
 	}
 	if err != nil {
 		log.Infof("Error fetching service %v from the store: %v", svcKey, err)
-		return nil
+		// Keep processing in case of service not found error
+		if !errors.IsNotFound(err) {
+			return nil, err
+		}
 	}
 	annotations := svc.(*v1.Service).Annotations
 	if _, ok := annotations[NPLSvcAnnotation]; !ok {
 		log.Errorf("NPL annotation %v not set on service %v", NPLSvcAnnotation, serviceName)
-		return nil
+		return nil, nil
 	}
 
 	selector := svc.(*v1.Service).Spec.Selector
 	if len(selector) == 0 {
 		log.Errorf("selector not set on service %v", serviceName)
-		return nil
+		return nil, nil
 	}
 	labelSelector, err := metav1.ParseToLabelSelector(labels.Set(selector).AsSelectorPreValidated().String())
 	labelmap, err := metav1.LabelSelectorAsMap(labelSelector)
 	if err != nil {
-		return nil
+		return nil, nil
 	}
 	pl, _ := createLabel(labels.SelectorFromSet(labelmap).String())
 	var podList []*v1.Pod
@@ -3523,9 +3550,9 @@ func (appMgr *Manager) GetPodsForService(namespace, serviceName string) []*v1.Po
 	}
 	if err != nil {
 		log.Debugf("Got error while listing Pods with selector %v: %v", selector, err)
-		return nil
+		return nil, err
 	}
-	return podList
+	return podList, nil
 }
 
 // getEndpointsForNPL returns members.

--- a/pkg/appmanager/appManager_test.go
+++ b/pkg/appmanager/appManager_test.go
@@ -1258,7 +1258,7 @@ var _ = Describe("AppManager Tests", func() {
 						Session: "user-enabled",
 					},
 				}
-				mems := mockMgr.appMgr.getEndpoints("cis.f5.com/as3-tenant=test", "default")
+				mems, _ := mockMgr.appMgr.getEndpoints("cis.f5.com/as3-tenant=test", "default")
 				Expect(mems).To(Equal(expMembers))
 			})
 
@@ -4120,11 +4120,11 @@ var _ = Describe("AppManager Tests", func() {
 				// no service and different namespace
 				mockMgr.appMgr.poolMemberType = NodePortLocal
 				mockMgr.appMgr.AddNamespace(namespace, selector, 0)
-				podList := mockMgr.appMgr.GetPodsForService("diff-ns", svcName)
+				podList, _ := mockMgr.appMgr.GetPodsForService("diff-ns", svcName)
 				Expect(podList).To(BeNil())
 
 				// no service created, correct namespace
-				podList = mockMgr.appMgr.GetPodsForService(namespace, svcName)
+				podList, _ = mockMgr.appMgr.GetPodsForService(namespace, svcName)
 				Expect(podList).To(BeNil())
 
 				//Add service, No nodeportlocal annotation
@@ -4135,21 +4135,21 @@ var _ = Describe("AppManager Tests", func() {
 				}
 				svc1 := test.NewService(svcName, "1", namespace, v1.ServiceTypeClusterIP, svcPorts)
 				appInf.svcInformer.GetStore().Add(svc1)
-				podList = mockMgr.appMgr.GetPodsForService(namespace, svcName)
+				podList, _ = mockMgr.appMgr.GetPodsForService(namespace, svcName)
 				Expect(podList).To(BeNil())
 
 				// Add nodeportlocal annotation
 				svc1.Annotations = make(map[string]string)
 				svc1.Annotations[NPLSvcAnnotation] = "true"
 				appInf.svcInformer.GetStore().Update(svc1)
-				podList = mockMgr.appMgr.GetPodsForService(namespace, svcName)
+				podList, _ = mockMgr.appMgr.GetPodsForService(namespace, svcName)
 				Expect(podList).To(BeNil())
 
 				// Add Pod selector
 				svc1.Spec.Selector = make(map[string]string)
 				svc1.Spec.Selector["app"] = "app1"
 				appInf.svcInformer.GetStore().Update(svc1)
-				podList = mockMgr.appMgr.GetPodsForService(namespace, svcName)
+				podList, _ = mockMgr.appMgr.GetPodsForService(namespace, svcName)
 				Expect(podList).To(BeNil())
 			})
 			It("Test getEndpoints", func() {
@@ -4170,7 +4170,7 @@ var _ = Describe("AppManager Tests", func() {
 					mockMgr.appMgr.isNodePort = true
 				}()
 				mockMgr.appMgr.AddNamespace(namespace, selector, 0)
-				members := mockMgr.appMgr.getEndpoints("test", namespace)
+				members, _ := mockMgr.appMgr.getEndpoints("test", namespace)
 				Expect(members).To(BeNil())
 
 				// Add service
@@ -4185,16 +4185,16 @@ var _ = Describe("AppManager Tests", func() {
 				svc2.Labels["test"] = "true"
 				mockMgr.appMgr.kubeClient.CoreV1().Services(namespace).Create(context.TODO(), svc1, metav1.CreateOptions{})
 				mockMgr.appMgr.kubeClient.CoreV1().Services(namespace).Create(context.TODO(), svc2, metav1.CreateOptions{})
-				members = mockMgr.appMgr.getEndpoints("test", namespace)
+				members, _ = mockMgr.appMgr.getEndpoints("test", namespace)
 				Expect(members).To(BeNil())
 
 				// Set isNodePort to false, no endpoints
 				mockMgr.appMgr.isNodePort = false
-				members = mockMgr.appMgr.getEndpoints("test", namespace)
+				members, _ = mockMgr.appMgr.getEndpoints("test", namespace)
 				Expect(members).To(BeNil())
 
 				// Set isNodePort to false, no endpoints
-				members = mockMgr.appMgr.getEndpoints("test", namespace)
+				members, _ = mockMgr.appMgr.getEndpoints("test", namespace)
 				Expect(members).To(BeNil())
 
 				// Add endpoints
@@ -4202,7 +4202,7 @@ var _ = Describe("AppManager Tests", func() {
 				endpts1 := test.NewEndpoints(svcName1, "1", "node0", namespace,
 					readyIps, nil, convertSvcPortsToEndpointPorts(svcPorts))
 				appInf.endptInformer.GetStore().Add(endpts1)
-				members = mockMgr.appMgr.getEndpoints("test", namespace)
+				members, _ = mockMgr.appMgr.getEndpoints("test", namespace)
 				Expect(members).NotTo(BeNil())
 				Expect(len(members)).To(Equal(2))
 
@@ -4345,14 +4345,14 @@ var _ = Describe("AppManager Tests", func() {
 				}
 
 				// Nodeport service
-				ret, err1, msg := mockMgr.appMgr.updatePoolMembersForNPL(svc, sKey, rsCfg, 0)
+				ret, err1, msg, _ := mockMgr.appMgr.updatePoolMembersForNPL(svc, sKey, rsCfg, 0)
 				Expect(ret).To(BeFalse())
 				Expect(len(err1)).NotTo(Equal(0))
 				Expect(len(msg)).NotTo(Equal(0))
 
 				// ClusterIP service not created
 				svc.Spec.Type = v1.ServiceTypeClusterIP
-				ret, err1, msg = mockMgr.appMgr.updatePoolMembersForNPL(svc, sKey, rsCfg, 0)
+				ret, err1, msg, _ = mockMgr.appMgr.updatePoolMembersForNPL(svc, sKey, rsCfg, 0)
 				Expect(ret).To(BeTrue())
 				Expect(len(err1)).To(Equal(0))
 				Expect(len(msg)).To(Equal(0))
@@ -4365,7 +4365,7 @@ var _ = Describe("AppManager Tests", func() {
 				svc.Spec.Selector = make(map[string]string)
 				svc.Spec.Selector["app"] = "app1"
 				appInf.svcInformer.GetStore().Add(svc)
-				ret, err1, msg = mockMgr.appMgr.updatePoolMembersForNPL(svc, sKey, rsCfg, 0)
+				ret, err1, msg, _ = mockMgr.appMgr.updatePoolMembersForNPL(svc, sKey, rsCfg, 0)
 				Expect(ret).To(BeTrue())
 				Expect(len(err1)).To(Equal(0))
 				Expect(len(msg)).To(Equal(0))

--- a/pkg/appmanager/ingress.go
+++ b/pkg/appmanager/ingress.go
@@ -1006,12 +1006,19 @@ func prepareV1IngressSSLContext(appMgr *Manager, ing *netv1.Ingress) {
 			continue
 		}
 		// Check if profile is contained in a Secret
-		secret, err := appMgr.kubeClient.CoreV1().Secrets(ing.ObjectMeta.Namespace).
-			Get(context.TODO(), tls.SecretName, metav1.GetOptions{})
-		if err != nil {
+		appInf, found := appMgr.getNamespaceInformer(ing.ObjectMeta.Namespace)
+		if !found {
+			log.Debugf("[CORE] No Informer found while fetching secret with name '%s' in namespace '%s'",
+				tls.SecretName, ing.ObjectMeta.Namespace)
+			continue
+		}
+		obj, found, err := appInf.secretInformer.GetIndexer().GetByKey(
+			fmt.Sprintf("%s/%s", ing.ObjectMeta.Namespace, tls.SecretName))
+		if err != nil || !found {
 			appMgr.rsrcSSLCtxt[tls.SecretName] = nil
 			continue
 		}
+		secret := obj.(*v1.Secret)
 		appMgr.rsrcSSLCtxt[tls.SecretName] = secret
 	}
 }

--- a/pkg/appmanager/ingress_test.go
+++ b/pkg/appmanager/ingress_test.go
@@ -127,7 +127,8 @@ var _ = Describe("V1 Ingress Tests", func() {
 					"tls.key": []byte("testkey"),
 				},
 			}
-			_, err := mockMgr.appMgr.kubeClient.CoreV1().Secrets(namespace).Create(context.TODO(), secret, metav1.CreateOptions{})
+			//_, err := mockMgr.appMgr.kubeClient.CoreV1().Secrets(namespace).Create(context.TODO(), secret, metav1.CreateOptions{})
+			err := mockMgr.appMgr.appInformers[namespace].secretInformer.GetStore().Add(secret)
 			Expect(err).To(BeNil())
 
 			spec := netv1.IngressSpec{

--- a/pkg/controller/nativeResourceWorker.go
+++ b/pkg/controller/nativeResourceWorker.go
@@ -848,9 +848,24 @@ func (ctlr *Controller) updatePoolMembersForRoutes(svc *v1.Service, updatePoolHe
 func (ctlr *Controller) processGlobalExtendedRouteConfig() {
 	splits := strings.Split(ctlr.routeSpecCMKey, "/")
 	ns, cmName := splits[0], splits[1]
-	cm, err := ctlr.kubeClient.CoreV1().ConfigMaps(ns).Get(context.TODO(), cmName, metav1.GetOptions{})
-	if err != nil {
+	var cm *v1.ConfigMap
+	var err error
+	var obj interface{}
+	var exist bool
+	nrInf, found := ctlr.getNamespacedNativeInformer(ns)
+	if found {
+		obj, exist, err = nrInf.cmInformer.GetIndexer().GetByKey(fmt.Sprintf("%s/%s", ns, cmName))
+		cm, _ = obj.(*v1.ConfigMap)
+	}
+	if !exist || cm == nil || err != nil {
+		// If informer fails to fetch configmap which may occur if cis just started which means informers may not have
+		// synced properly then try to fetch using kubeClient
+		cm, err = ctlr.kubeClient.CoreV1().ConfigMaps(ns).Get(context.TODO(), cmName, metav1.GetOptions{})
+	}
+	// Skip processing further if Extended configmap is not found
+	if err != nil || cm == nil {
 		log.Errorf("Unable to Get Extended Route Spec Config Map: %v, %v", ctlr.routeSpecCMKey, err)
+		return
 	}
 	err = ctlr.setNamespaceLabelMode(cm)
 	if err != nil {

--- a/pkg/resource/resourceConfig.go
+++ b/pkg/resource/resourceConfig.go
@@ -2754,7 +2754,7 @@ func CreatePolicy(rls Rules, policyName, partition string) *Policy {
 	return &plcy
 }
 
-func (cm *AgentCfgMap) Init(n string, ns string, d string, l map[string]string, getEP func(string, string) []Member) {
+func (cm *AgentCfgMap) Init(n string, ns string, d string, l map[string]string, getEP func(string, string) ([]Member, error)) {
 	cm.Name = n
 	cm.Namespace = ns
 	cm.Data = d

--- a/pkg/resource/resourceConfig_test.go
+++ b/pkg/resource/resourceConfig_test.go
@@ -2215,7 +2215,7 @@ var _ = Describe("Resource Config Tests", func() {
 		It("Test AgentCfgMap Init", func() {
 			cm := &AgentCfgMap{}
 			label := make(map[string]string)
-			getEP := func(string, string) []Member { return nil }
+			getEP := func(string, string) ([]Member, error) { return nil, nil }
 			cm.Init("test1", "default", "", label, getEP)
 			Expect(cm.Name).To(Equal("test1"))
 			Expect(cm.Namespace).To(Equal("default"))

--- a/pkg/resource/types.go
+++ b/pkg/resource/types.go
@@ -415,7 +415,7 @@ type (
 
 	AgentCfgMap struct {
 		Operation    string
-		GetEndpoints func(string, string) []Member
+		GetEndpoints func(string, string) ([]Member, error)
 		Data         string
 		Name         string
 		Namespace    string

